src/target/riscv/riscv.c

```diff
@@ -5043,89 +5121,54 @@ int riscv_enumerate_triggers(struct target *target)
 	if (r->triggers_enumerated)
 		return ERROR_OK;
 
-	r->triggers_enumerated = true;	/* At the very least we tried. */
+	if (target->state != TARGET_HALTED) {
+		LOG_TARGET_ERROR(target, "Unable to enumerate triggers: target not halted.");
+		return ERROR_FAIL;
+	}
 
-	riscv_reg_t tselect;
-	int result = riscv_get_register(target, &tselect, GDB_REGNO_TSELECT);
+	riscv_reg_t orig_tselect;
+	int result = riscv_get_register(target, &orig_tselect, GDB_REGNO_TSELECT);
 	/* If tselect is not readable, the trigger module is likely not
-		* implemented. There are no triggers to enumerate then and no error
-		* should be thrown. */
+	 * implemented. */
 	if (result != ERROR_OK) {
-		LOG_TARGET_DEBUG(target, "Cannot access tselect register. "
-				"Assuming that triggers are not implemented.");
-		r->trigger_count = 0;
-		return ERROR_OK;
+		LOG_DEBUG("[%s] Cannot access tselect register. "
+				"Assuming that triggers are not implemented.", target_name(target));
+		return ERROR_TARGET_RESOURCE_NOT_AVAILABLE;
 	}
 
-	for (unsigned int t = 0; t < RISCV_MAX_TRIGGERS; ++t) {
-		r->trigger_count = t;
-
-		/* If we can't write tselect, then this hart does not support triggers. */
-		if (riscv_set_register(target, GDB_REGNO_TSELECT, t) != ERROR_OK)
+	unsigned int t = 0;
+	unsigned int trigger_tinfo[ARRAY_SIZE(r->trigger_tinfo)];
+	for (; t < ARRAY_SIZE(r->trigger_tinfo); ++t) {
+		result = check_if_trigger_is_availible(target, t);
+		if (result == ERROR_FAIL)
+			return ERROR_FAIL;
```

Comment:
Should we try to restore the `orig_tselect` even if the enumeration fails? I believe so.

In that case, you can use the `goto fail;` pattern instead of direct `return ERROR_FAIL`.

Reply:
`riscv_enumerate_triggers` returns `ERROR_FAIL` in case `tselect` or `tdata1` is not accessible. It looks like something is seriously wrong with trigger module and there is no benefit in attempting a restore of `tselect`, unless I'm missing something.

Reply:
Ok, agreed. Let's leave the current implementation, then.

Reply:
Hi, I noticed this:
check_if_trigger_is_availible
which should be:
check_if_trigger_is_available

Reply:
> Hi, I noticed this: check_if_trigger_is_availible which should be: check_if_trigger_is_available

Addressed.